src/components/interface/cbuf_mgr/cos_alloc.c

```diff
@@ -0,0 +1,536 @@
+#if ALLOC_DEBUG >= ALLOC_DEBUG_ALL
+			if (alloc_debug) printc("malloc (in %d): returning memory region @ %x, size %d, index %d, (head of list is %x)", 
+						cos_spd_id(), ptr, size, idx, __small_mem[idx]);
+#endif
+			return start;
+		} 
+		next = ptr->next;
+		//__small_mem[idx]=ptr->next;
+	} while (unlikely(cos_cmpxchg(&__small_mem[idx], (long)ptr, (long)next) != (long)next));
+	ptr->next=0;
+
+#if ALLOC_DEBUG >= ALLOC_DEBUG_ALL
+	if (alloc_debug) printc("malloc (in %d): returning memory region @ %x, size %d, index %d, (head of list is %x)", 
+				cos_spd_id(), ptr, size, idx, __small_mem[idx]);
+#endif
+	/* 
+	 * FIXME: This still suffers from the ABA problem -- still a
+	 * race if ptr and next are removed from list, then ptr put
+	 * back in between ptr=__small_mem[idx] and the cmpxchg.  next
+	 * is used elsewhere, but we are still going to put it back at
+	 * the head of the list.  This can be solved with another
+	 * cos_cmpxchg loop to verify next as well.
+	 */
+	alloc_stats_report(DBG_ALLOC, idx);
+
+	return ptr;
+}
+
+/* -- PUBLIC FUNCTIONS ---------------------------------------------------- */
+
+static void do_cbuf_free(cbuf_t cbid)
+{
+    printc("        calling cbuf free\n");
+    cbuf_free(cbid);
+}
+
+static void _alloc_libc_free(void *ptr) 
+{
+    printc("    calling free\n");
+    if (ptr)
+    {
+        printc("    found a pointer\n");
+        __alloc_t *pointer = BLOCK_START(ptr);
+        size_t size = pointer->size;
+
+        printc("    size is %zu\n", size);
+        if (size)
+        {
+            cbuf_free(pointer->cbuf_id);
+        }
+    }
+}
+
+void free(void *ptr) __attribute__((weak,alias("_alloc_libc_free")));
+
+#ifdef WANT_MALLOC_ZERO
+static __alloc_t zeromem[2];
+#endif
+
+static void* _alloc_libc_malloc(size_t size) {
+  __alloc_t* ptr;
+  size_t need;
+#ifdef WANT_MALLOC_ZERO
+  if (!size) return BLOCK_RET(zeromem);
+#else
+  if (!size) goto err_out;
+#endif
+  size+=sizeof(__alloc_t);
+  if (unlikely(size<sizeof(__alloc_t))) goto err_out;
+
+  if (size<=__MAX_SMALL_SIZE) {
+    need=GET_SIZE(size);
+    ptr=__small_malloc(need);
+  } else {
+    need=PAGE_ALIGN(size);
+    cbuf_t cbid;
+    ptr = need ? do_cbuf_alloc(need, &cbid) : MAP_FAILED;
+    ptr->cbuf_id = cbid;
+  }
+  if (ptr==MAP_FAILED) goto err_out;
+  ptr->size=need;
+  return BLOCK_RET(ptr);
+err_out:
+  printc("\treturning malloc error\n");
+  return 0;
+}
+
+void* malloc(size_t size) __attribute__((alias("_alloc_libc_malloc")));
+
+void *__libc_calloc(size_t nmemb, size_t _size)
+{
+	size_t tot = nmemb*_size;
+	char *ret = malloc(tot);
+	
+	memset(ret, 0, tot);
+	return ret;
+}
+
+void* calloc(size_t nmemb, size_t _size) __attribute__((weak,alias("__libc_calloc")));
+
+/* gabep1 additions for allocations of pages. */
+
+void *alloc_page(void)
+{
+	struct free_page *fp;
+	void *a;
+
+	fp = page_list.next;
+	if (NULL == fp) {
+        cbuf_t cbid;
+		a = do_cbuf_alloc(PAGE_SIZE, &cbid);
+        // what if this fails?
```

Comment:
`return NULL`  In systems code, you always have to check the return value of memory allocation calls, and deal with the contingencies.